hare/eligibility/oracle.go

```diff
@@ -398,58 +418,92 @@ func (o *Oracle) actives(layer types.LayerID) (map[string]uint64, error) {
 	o.lock.Lock()
 	defer o.lock.Unlock()
 
-	// check cache
-	if val, exist := o.activesCache.Get(safeEp); exist {
-		return val.(map[string]uint64), nil
+	// check cache first
+	if val, exist := o.activesCache.Get(targetLayer.GetEpoch()); exist {
```

Comment:
This is problematic. The target layer's epoch alone doesn't determine the safe layers, which determines the number of activations. If the target layer is before the epoch offset, the previous epoch is used (see lines 88-94 in this file).

I suggest we remove lines 88-94 and replace them with a method that determines in which epoch the safe layers should be - then we can use that function to determine the cache key, as well.

Reply:
> replace them with a method that determines in which epoch the safe layers should be

This is exactly what `safeLayerRange()` already does, isn't it? The answer is, `safeLayerStart.GetEpoch()` and `safeLayerEnd.GetEpoch()`.

> we can use that function to determine the cache key, as well

I don't think this is necessary. Think of the cache as memoizing this method call. If the method is called twice with the same input, it should return the same output. The cache key does _not_ need to agree with the safe layer.

Reply:
Let's take an example. Say we have 10 layer epochs and 13 layer safety param, epoch offset is 3.

`targetLayer = 44`. We get `safeLayerStart = 20, safeLayerEnd = 23`. We cache the result for `targetLayer.GetEpoch() = 4`.

Then we get to `targetLayer = 47`. We get a cache hit for `targetLayer.GetEpoch() = 4` which was calculated with `safeLayerStart = 20, safeLayerEnd = 23` but the real result we want should be calculated with `safeLayerStart = 30, safeLayerEnd = 33`.

Added comments with actual values to the code I'm referring to:
```golang
	safeLayer := targetLayer - safetyParam               // safeLayer = 44 - 13 = 31   | 47 - 13 = 34
	safeEpoch := safeLayer.GetEpoch()                    // safeEpoch = 3              | 3
	safeLayerStart = safeEpoch.FirstLayer()              // safeLayerStart = 30        | 30
	safeLayerEnd = safeLayerStart + epochOffset          // safeLayerEnd = 30 + 3 = 33 | 33

	// If the safe layer is in the first epochOffset layers of an epoch,
	// return a range from the beginning of the previous epoch
	if safeLayer < safeLayerEnd {                        // 31 < 33 => true            | 34 < 33 => false
		// prevent overflow
		if safeLayerStart <= layersPerEpoch {
			return types.GetEffectiveGenesis(), types.GetEffectiveGenesis()
		}
		safeLayerStart = safeLayerStart - layersPerEpoch // safeLayerStart = 30 - 10 = 20
		safeLayerEnd = safeLayerEnd - layersPerEpoch     // safeLayerEnd = 33 - 10 = 23
	}
```

Reply:
thanks for the clear example. fixed this. please have another look!